### ISSUE TITLE
TransactionException: Lock is not owned by the transaction!

### ISSUE THREAD

**Issue by pveentjer** [hazelcast/hazelcast#6039]
Problem with transactions and locking. 

This is a problem with 3.5 and 3.6.

Will provide more details later

```
com.hazelcast.transaction.TransactionException: Lock is not owned by the transaction! Caller: ad150474-bc18-470f-82d6-4e5a7339ab44, Owner: Owner: ad150474-bc18-470f-82d6-4e5a7339ab44, thread-id: 251
	at com.hazelcast.map.impl.tx.TxnPrepareBackupOperation.run(TxnPrepareBackupOperation.java:49)
	at com.hazelcast.spi.impl.operationservice.impl.operations.Backup.run(Backup.java:102)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:160)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:349)
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.processPacket(OperationThread.java:166)
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.process(OperationThread.java:135)
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.doRun(OperationThread.java:122)
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.run(OperationThread.java:97)
```

**Comment by pveentjer** [comment]
The following simulator tests reproduces the problem:

```
package com.hazelcast.simulator.tests.map;

import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.TransactionalMap;
import com.hazelcast.logging.ILogger;
import com.hazelcast.logging.Logger;
import com.hazelcast.simulator.test.TestContext;
import com.hazelcast.simulator.test.annotations.RunWithWorker;
import com.hazelcast.simulator.test.annotations.Setup;
import com.hazelcast.simulator.worker.tasks.AbstractMonotonicWorker;
import com.hazelcast.transaction.TransactionContext;
import com.hazelcast.transaction.TransactionOptions;
import com.hazelcast.transaction.TransactionOptions.TransactionType;

import java.io.Externalizable;
import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectOutput;
import java.util.concurrent.ThreadLocalRandom;

public class MapTransactionContextTest {

    private static final ILogger LOGGER = Logger.getLogger(MapTransactionContextTest.class);

    // properties
    public TransactionType transactionType = TransactionType.TWO_PHASE;
    public int durability = 1;
    public int range = 1000;

    private HazelcastInstance hz;

    @Setup
    public void setup(TestContext testContext) throws Exception {
        hz = testContext.getTargetInstance();
    }

    @RunWithWorker
    public Worker createWorker() {
        return new Worker();
    }

    private class Worker extends AbstractMonotonicWorker {

        @Override
        protected void timeStep() {
            int key = nextRandom(0, range / 2);

            TransactionOptions transactionOptions = new TransactionOptions().setTransactionType(transactionType).setDurability(durability);

            TransactionContext transactionContext = hz.newTransactionContext(transactionOptions);

            transactionContext.beginTransaction();

            TransactionalMap<Object, Object> txMap = transactionContext.getMap("map");

            try {
                Object val = txMap.getForUpdate(key);

                if (val != null) {
                    key = nextRandom(range / 2, range);
                }

                txMap.put(key, new SampleValue(key));

                transactionContext.commitTransaction();
            } catch (Exception e) {

                e.printStackTrace();

                transactionContext.rollbackTransaction();
            }
        }

        protected int nextRandom(int min, int max) {
            return ThreadLocalRandom.current().nextInt(max - min) + min;
        }
    }

    /**
     * Entity class for benchmark.
     */
    public static class SampleValue implements Externalizable {
        /** */
        private int id;

        /** */
        public SampleValue() {
            // No-op.
        }

        /**
         * @param id Id.
         */
        public SampleValue(int id) {
            this.id = id;
        }

        /**
         * @return Id.
         */
        public int id() {
            return id;
        }

        /**
         * {@inheritDoc}
         */
        @Override
        public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
            id = in.readInt();
        }

        /**
         * {@inheritDoc}
         */
        @Override
        public void writeExternal(ObjectOutput out) throws IOException {
            out.writeInt(id);
        }

        /**
         * {@inheritDoc}
         */
        @Override
        public String toString() {
            return "Value [id=" + id + ']';
        }
    }
}
````

**Comment by pveentjer** [comment]
When I switch to executeTransaction, the problems remains

**Comment by pveentjer** [comment]
The problem is independent of 'durability' setting.

**Comment by pveentjer** [comment]
More logging. I have included the thread in the logging that is passsed into the TxnPrepareBackupOperation:

```
com.hazelcast.transaction.TransactionException: ---------------------Lock is not owned by the transaction! Caller: 7898fce3-9cce-44e8-8119-54a66786cf3c lockThreadId: 117, Owner: Owner: cfd0bac4-42a7-4953-9233-c11a4c973abc, thread-id: 109
        at com.hazelcast.map.impl.tx.TxnPrepareBackupOperation.run(TxnPrepareBackupOperation.java:50)
        at com.hazelcast.spi.impl.operationservice.impl.operations.Backup.run(Backup.java:102)
        at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:160)
        at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:349)
        at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.processPacket(OperationThread.java:166)
        at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.process(OperationThread.java:135)
        at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.doRun(OperationThread.java:122)
        at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.run(OperationThread.java:97)
```

**Comment by pveentjer** [comment]
My impression is that the backup is not in sync with the primary. The prepare on the primary succeeds, but the prepare on the backup fails because it is out of sync. This causes an exception to be thrown during the processing of this backup operation, but this exception is swallowed. So the invocation has no clue what has gone wrong.

Eventually the invocation will stop waiting on the backup and then continue. 

That is why I'm not seeing any failures in the simulator; but I see the exception in the logging. And because of the waiting for the timeout on the failed backup, the performance is very bad.

**Comment by pveentjer** [comment]
I think the following flow explains the issue. Need to get it confirmed:

t1: unlock primary (completing its transaction)
t2: lock primary (call TxnLockAndGetOperation)--> this call is not replicated to the backup
t2: lock primary (TxnPrepareOperation)
t2: lock backup  (TxnPrepareBackupOperation) <--- crash and burn because the unlock of the backup by t1 has not completed.
t1: unlock backup

**Comment by bwzhang2011** [comment]
@pveentjer, any update with such issue ? as you pointed out, the problem followed up with hz3.5+ but not with hz3.4+, was it right ?

**Comment by pveentjer** [comment]
I don't know if the problem exists with 3.4; have not checked it. I have a 'temp' solution and performance problem is completely resolved,  but it suppresses the exception..  it doesn't solve the reason the exception happened. I hope we can provide a solution shortly so it can be fixed in 3.5.3

**Comment by pveentjer** [comment]
The following unit test reproduces it in seconds

```
package com.hazelcast.map;

import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.TransactionalMap;
import com.hazelcast.test.HazelcastTestSupport;
import com.hazelcast.test.TestThread;
import com.hazelcast.transaction.TransactionContext;
import com.hazelcast.transaction.TransactionOptions;
import org.junit.Before;
import org.junit.Test;

import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicBoolean;

public class MapTxFailureTest extends HazelcastTestSupport {
    private AtomicBoolean stop = new AtomicBoolean();
    // properties
    public TransactionOptions.TransactionType transactionType = TransactionOptions.TransactionType.TWO_PHASE;
    public int durability = 1;
    public int range = 1000;

    private HazelcastInstance[] cluster;
    private int workerCount = 5;
    private Worker[] workers;

    @Before
    public void setup() {
        setLoggingLog4j();
        cluster = createHazelcastInstanceFactory(2).newInstances();

        workers = new Worker[workerCount];
        for (int k = 0; k < workerCount; k++) {
            workers[k] = new Worker();
        }
    }

    @Test
    public void test() {
        for (Worker worker : workers) {
            worker.start();
        }

        sleepAndStop(stop, 60);

        for (Worker worker : workers) {
            worker.assertSucceedsEventually();
        }
    }

    private class Worker extends TestThread {


        public void doRun() {
            while (!stop.get()) {
                timeStep();
            }
        }

        @Override
        public void onError(Throwable t) {
            stop.set(true);
        }

        protected void timeStep() {
            int key = nextRandom(0, range / 2);

            TransactionOptions transactionOptions = new TransactionOptions().setTransactionType(transactionType).setDurability(durability);

            TransactionContext transactionContext = getTargetInstance().newTransactionContext(transactionOptions);

            transactionContext.beginTransaction();

            TransactionalMap<Object, Object> txMap = transactionContext.getMap("map");

            try {
                Object val = txMap.getForUpdate(key);

                if (val != null) {
                    key = nextRandom(range / 2, range);
                }

                txMap.put(key, new Long(key));

                transactionContext.commitTransaction();
            } catch (Exception e) {

                e.printStackTrace();

                transactionContext.rollbackTransaction();
            }
        }

        private HazelcastInstance getTargetInstance() {
            int index = ThreadLocalRandom.current().nextInt(cluster.length);
            return cluster[index];
        }

        protected int nextRandom(int min, int max) {
            return ThreadLocalRandom.current().nextInt(max - min) + min;
        }
    }
}
```
